File: dynamic-programming/best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.ts
function maxProfit(prices: number[], fee: number): number {
  const n: number = prices.length;
  let profits = 0;
  let hold: number = -prices[0];

  for (let i = 1; i < n; i++) {
    profits = Math.max(profits, hold + prices[i] - fee);
    hold = Math.max(hold, profits - prices[i]);
  }
  return profits;
}

File: dynamic-programming/best-time-to-buy-and-sell-stock-with-transaction-fee/best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution {
  public int maxProfit(int[] prices, int fee) {
    int n = prices.length;
    int profits = 0;
    int hold = -prices[0];

    for(int i = 1; i < n; i++) {
      profits = Math.max(profits, hold + prices[i] - fee);
      hold = Math.max(hold, profits - prices[i]);
    }

    return profits;
  }
}
